lib/publiccloud/utils.pm

```diff
@@ -885,4 +887,129 @@ sub upload_asset_on_remote {
+    my $zcmd = $args{cmd};
+    # full command to run in ssh
+    $args{cmd} = "sudo zypper -n $zcmd";
+    #
+    delete $args{exitcode};
+    delete $args{retry};
+    delete $args{delay};
+    # retry loop
+    my ($ret, $exit);
+    for (1 .. $retry) {
+        # pause on next
+        sleep($delay) if (defined($ret));
+        # remote execution
+        $exit = $instance->run_ssh_command(%args);
+        # to avoid eventual 'undef' errors in expressions:
+        $ret = $exit // -1;
+        last if ($ret == 0);
+        # check exit codes
+        if ($ret == 4) {
+            if ($instance->ssh_script_run(qq[sudo grep "Error code.*502" $log]) == 0) {
+                die 'According to bsc#1070851 zypper should automatically retry internally. Bugfix missing for current product?';
+            }
+            elsif ($instance->ssh_script_run(qq[sudo grep "Solverrun finished with an ERROR" $log] == 0)) {
+                my $search_conflicts = q[sudo awk 'BEGIN {print "Processing conflicts - ",NR; group=0}
+                    /Solverrun finished with an ERROR/,/statistics/{ 
+                    print group"|", $0; if ($0 ~ /statistics/ ){ print "EOL"; group++ }; }' ] . $log;
+                my $conflicts = $instance->ssh_script_output($search_conflicts);
+                record_info("Conflict", $conflicts, result => 'fail');
+                diag "Package conflicts found, not retrying anymore" if $conflicts;
+                last;
+            }
+            next;
+        }
+        last;
+    }
+    # Store 'install' / 'remove' actions for later use by tests/console/zypper_log_packages.pm
+    zyp_pkg_store($zcmd, $exit);
+    # Result management
+    my $command = $args{cmd};
+    unless (grep { $_ == $ret } @$exit_codes) {
+        $instance->upload_log($log);
+        my $msg = qq[$command failed with code: $exit];
+        if ($ret == 104) {
+            $msg .= " (ZYPPER_EXIT_INF_CAP_NOT_FOUND)\n\nRelated zypper logs:\n";
+            $instance->ssh_script_run(qq[sudo tac $log | grep -F -m1 -B100000 "Hi, me zypper" | tac | grep -E '(SolverRequester.cc|THROW|CAUGHT)' > /tmp/z104.txt]);
+            $msg .= $instance->ssh_script_output('cat /tmp/z104.txt');
+        }
+        elsif ($ret == 107) {
+            $msg .= " (ZYPPER_EXIT_INF_RPM_SCRIPT_FAILED)\n\nRelated zypper logs:\n";
+            $instance->ssh_script_run(qq[sudo tac $log | grep -F -m1 -B100000 "Hi, me zypper" | tac | grep -E 'RpmPostTransCollector.cc(executeScripts):.* scriptlet failed, exit status' > /tmp/z107.txt]);
+            $msg .= $instance->ssh_script_output('cat /tmp/z107.txt') . "\n\n";
+        }
+        else {
+            $instance->ssh_script_run(qq[sudo tac $log | grep -F -m1 -B100000 "Hi, me zypper" | tac | grep 'Exception.cc' > /tmp/zlog.txt]);
+            $msg .= "\n\nRelated zypper logs:\n";
+            $msg .= $instance->ssh_script_output('cat /tmp/zlog.txt');
+        }
+        die $msg unless ($proceed);
```

Comment:
I don't think we need to allow caller to ignore zypper failures . caller already have ability to ignore certain exit codes . I think it is more than enough . I can hardly imagine case when you can just fully ignore whatever zypper do 🤔